src/interface/update.hpp

```diff
@@ -13,44 +13,114 @@
 #ifndef INTERFACE_UPDATE_HPP_
 #define INTERFACE_UPDATE_HPP_
 
+#include <algorithm>
+#include <functional>
+#include <limits>
 #include <memory>
 #include <string>
+#include <typeinfo>
+#include <utility>
 #include <vector>
 
 #include "defs.hpp"
-#include "interface/mesh_data.hpp"
-#include "interface/meshblock_data.hpp"
-#include "mesh/mesh.hpp"
+#include "interface/metadata.hpp"
+#include "interface/params.hpp"
+#include "interface/state_descriptor.hpp"
 
 #include "kokkos_abstraction.hpp"
 
 namespace parthenon {
 
 namespace Update {
 
-TaskStatus FluxDivergenceBlock(std::shared_ptr<MeshBlockData<Real>> &in,
-                               std::shared_ptr<MeshBlockData<Real>> &dudt_cont);
-TaskStatus FluxDivergenceMesh(std::shared_ptr<MeshData<Real>> &in_pack,
-                              std::shared_ptr<MeshData<Real>> &dudt_pack);
-void UpdateMeshBlockData(std::shared_ptr<MeshBlockData<Real>> &in,
-                         std::shared_ptr<MeshBlockData<Real>> &dudt_cont, const Real dt,
-                         std::shared_ptr<MeshBlockData<Real>> &out);
-void UpdateMeshData(std::shared_ptr<MeshData<Real>> &in,
-                    std::shared_ptr<MeshData<Real>> &dudt, const Real dt,
-                    std::shared_ptr<MeshData<Real>> &out);
-void AverageMeshData(std::shared_ptr<MeshData<Real>> &c1_pack,
-                     std::shared_ptr<MeshData<Real>> &c2_pack, const Real wgt1);
-Real EstimateTimestep(std::shared_ptr<MeshBlockData<Real>> &rc);
+template <typename T>
+TaskStatus FluxDivergence(std::shared_ptr<T> &in, std::shared_ptr<T> &dudt_obj);
 
+template <typename T>
+void UpdateIndependentData(T &in, T &dudt, const Real dt, T &out) {
+  std::vector<MetadataFlag> flags({Metadata::Independent});
+  auto in_pack = in->PackVariables(flags);
+  auto out_pack = out->PackVariables(flags);
+  auto dudt_pack = dudt->PackVariables(flags);
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "UpdateMeshData", DevExecSpace(), 0, in_pack.GetDim(5) - 1, 0,
+      in_pack.GetDim(4) - 1, 0, in_pack.GetDim(3) - 1, 0, in_pack.GetDim(2) - 1, 0,
+      in_pack.GetDim(1) - 1,
+      KOKKOS_LAMBDA(const int b, const int l, const int k, const int j, const int i) {
+        out_pack(b, l, k, j, i) = in_pack(b, l, k, j, i) + dt * dudt_pack(b, l, k, j, i);
+      });
+}
```

Comment:
You could put the implementations of these functions in the `update.cpp` file if you explicitly instantiated them for `MeshData` and `MeshBlockData` at the bottom of the file.